go/libkb/pseudonym.go

```diff
@@ -0,0 +1,262 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkb
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/keybase/go-codec/codec"
+
+	context "golang.org/x/net/context"
+)
+
+// TLFPseudonym is an identifier for a key in a tlf
+type TlfPseudonym [32]byte
+
+type keyGen int
+type tlfID [16]byte
+
+const tlfPseudonymVersion = 1
+
+// TlfPseudonymInfo is what a pseudonym represents
+type TlfPseudonymInfo struct {
+	// TLF name like: /keybase/private/a,b
+	Name string
+	// TLF id
+	ID      tlfID
+	KeyGen  keyGen
+	HmacKey [32]byte
+}
+
+func (t TlfPseudonym) String() string {
+	return hex.EncodeToString(t[:])
+}
+
+func (t *TlfPseudonym) Eq(r *TlfPseudonym) bool {
+	if t != nil && r != nil {
+		return bytes.Equal(t[:], r[:])
+	}
+	return (t == nil) && (r == nil)
+}
+
+// tlfPseudonymReq is what the server needs to store a pseudonym
+type tlfPseudonymReq struct {
+	Pseudonym string `json:"tlf_pseudonym"` // hex
+	Name      string `json:"tlf_name"`
+	ID        string `json:"tlf_id"` // hex
+	KeyGen    int    `json:"tlf_key_gen"`
+	HmacKey   string `json:"hmac_key"` // hex
+}
+
+// tlfPseudonymContents is the data packed inside the HMAC
+type tlfPseudonymContents struct {
+	_struct bool `codec:",toarray"`
+	Version int
+	Name    string
+	ID      tlfID
+	KeyGen  keyGen
+}
+
+type getTlfPseudonymsRes struct {
+	TlfPseudonyms []getTlfPseudonymRes `json:"tlf_pseudonyms"`
+	Status        AppStatus            `json:"status"`
+}
+
+func (r *getTlfPseudonymsRes) GetAppStatus() *AppStatus {
+	return &r.Status
+}
+
+type getTlfPseudonymRes struct {
+	Err  *PseudonymGetError `json:"err"`
+	Info *struct {
+		Name    string `json:"tlf_name"`
+		ID      string `json:"tlf_id"` // hex
+		KeyGen  int    `json:"tlf_key_gen"`
+		HmacKey string `json:"hmac_key"` // hex
+	} `json:"info"`
+}
+
+type GetTlfPseudonymEither struct {
+	// Exactly one of these 2 fields is nil.
+	Err  error
+	Info *TlfPseudonymInfo
+}
+
+// MakePseudonym makes a TLF pseudonym from the given input.
+func MakePseudonym(info TlfPseudonymInfo) (TlfPseudonym, error) {
+	input := tlfPseudonymContents{
+		Version: tlfPseudonymVersion,
+		Name:    info.Name,
+		ID:      info.ID,
+		KeyGen:  info.KeyGen,
+	}
+	mh := codec.MsgpackHandle{WriteExt: true}
+	var buf []byte
+	enc := codec.NewEncoderBytes(&buf, &mh)
+	err := enc.Encode(input)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	mac := hmac.New(sha256.New, info.HmacKey[:])
+	mac.Write(buf)
+	hmacArr := mac.Sum(nil)
+	var hmac [32]byte
+	if len(hmacArr) != len(hmac) {
+		return [32]byte{}, fmt.Errorf(
+			"Expected array of size %d, got %d",
+			len(hmac), len(hmacArr))
+	}
+	copy(hmac[:], hmacArr)
```

Comment:
We do this copy dance in a bunch of different places. I wonder if it would be worth having some generic functions for it in the most common sizes. Maybe err-returning versions and panicking versions? Then the caller could just cast the result and be done. More of a thought for the future.